ANALYSIS/ANALYSIS/AliAnalysisManager.cxx

```diff
@@ -3139,3 +3148,92 @@ void AliAnalysisManager::InitInputData(AliVEvent* esdEvent, AliVfriendEvent* esd
+ * @param iStartWithFile starting from file 'iStartWithFile' (>= 1).
+ * @param cTreeNameArg Tree name for chaining. if "auto" it will be taken as the first tree name from the first file from filelist
+ * @param friends Specify the root_file/friend_tree that is assumed to be in the same directory as the each input file; if friend_tree is not specified we will assume the defaults
+ * @return TChain*
+ */
+TChain* AliAnalysisManager::CreateChain(const char* filelist, const char* cTreeNameArg, const char* friends, Int_t iNumFiles, Int_t iStartWithFile)
+{
+TString sTreeNameArg (cTreeNameArg), treeName;
+
+TFileCollection filecoll ("anachain","File collection for analysis"); // easy manipulation of file collections
+Int_t iAddedFiles = filecoll.AddFromFile(filelist,iNumFiles,iStartWithFile);
+if ( iAddedFiles < 1 ) { std::cout << "NO Files added to collection !!!" << std::endl; return NULL; }
+
+// if cTreeNameArg is auto lets try to autodetect what type of tree we have;
+// the assuption is that all files are the same and the first one is reprezentative
+THashList* list =  filecoll.GetList();
+if ( sTreeNameArg.EqualTo("auto") ) { // if tree name is not specified
+  TRegexp tree_regex ("[aod,esd]Tree");
+  TFileInfo* fileinfo = dynamic_cast<TFileInfo*>(list->At(0)); // get first fileinfo in list
+  TFile file (fileinfo->GetFirstUrl()->GetFile()); // get the actual TFile
+  if (file.IsZombie()) { cout << "Should not reach this message!! Error opening file" << endl; return NULL; }
+
+  // lets parse the TFile
+  TIter next(file.GetListOfKeys());
+  TKey* key = NULL;
+  while (( key = dynamic_cast<TKey*>(next()) )) {
+    TString class_name = key->GetClassName();
+    if ( ! class_name.EqualTo("TTree") ) { continue; } // searching for first TTree
+
+    TString key_name = key->GetName();
+    if ( key_name.Contains(tree_regex) ) { treeName = key_name; break;} // that is named either aodTree or esdTree
+    }
+  file.Close();
+  }
+else
+  { treeName = sTreeNameArg ; } // tree name is specified
+
+TChain* chain = new TChain (treeName.Data(),""); // lets create the chain
+if ( chain->AddFileInfoList(list) == 0 ) { return NULL; } // and add file collection (THashList is a Tlist that is a TCollection)
+
+// start treatment of friends
+TChain* chainFriend = NULL;
+TString sFriends (friends);
+if (!sFriends.IsNull()) {
+  TString friends_treename, friends_filename;
+  TObjArray* arr = sFriends.Tokenize("/");
+  TObjString* strobj_file = dynamic_cast<TObjString*>(arr->At(0));
+  if (strobj_file) { friends_filename = strobj_file->GetString(); }
+  TObjString* strobj_tree = dynamic_cast<TObjString*>(arr->At(1));
+  if (strobj_tree) { friends_treename = strobj_tree->GetString(); }
+  delete arr;
+
+  if (friends_treename.IsNull()) {
+    if (treeName.EqualTo("esdTree")) {
+      friends_treename = "esdFriendTree"; }
+    else if (treeName.EqualTo("aodTree")) {
+      friends_treename = "aodTree"; }
```

Comment:
What means aodTree as a friend for the aodTree?

Reply:
i have no idea ... the logic is from https://github.com/alisw/AliPhysics/blob/master/PWG/EMCAL/macros/CreateAODChain.C#L38